### PR TITLE
feat(gatsby-plugin-utils): add package and methods to validate plugin options

### PR DESCRIPTION
**Pull request by mxstbr**
This adds a new package to the monorepo (`gatsby-plugin-utils`) that currently exposes a single method:

```JS
import { validateOptionsSchema } from "gatsby-plugin-utils"

const validatedOptions = await validateOptionsSchema(pluginSchema, pluginOptions);

// Ignore "external" (i.e. async) validation rules
const validatedOptions = await validateOptionsSchema(pluginSchema, pluginOptions, {
  validateExternalRules: false
]);
```

[ch15944]

### CODE REVIEW

File: packages/gatsby-plugin-google-analytics/package.json
Comment (ascorbic) [inline packages/gatsby-plugin-google-analytics/package.json:47]: Did you mean to remove this, if we're using Joi?

File: packages/gatsby-plugin-google-analytics/package.json
Comment (wardpeet) [inline packages/gatsby-plugin-google-analytics/package.json:106]: Is this intentional?

File: packages/gatsby-plugin-utils/package.json
Comment (ascorbic) [inline packages/gatsby-plugin-utils/package.json:8]: Can we built types too

File: packages/gatsby-plugin-utils/src/__tests__/index.ts
Comment (ascorbic) [inline packages/gatsby-plugin-utils/src/__tests__/index.ts:57]: Rather than a snapshot, a more explicit check on it not throwing would be better

File: packages/gatsby-plugin-utils/src/validate.ts
Comment (ascorbic) [inline packages/gatsby-plugin-utils/src/validate.ts:22]: Are we able to guarantee that there will never be any async rules, if they're not using `external`?

File: packages/gatsby-plugin-utils/src/validate.ts
Comment (ascorbic) [inline packages/gatsby-plugin-utils/src/validate.ts:30]: No need to await a return value

File: packages/gatsby-plugin-utils/src/validate.ts
Comment (ascorbic) [inline packages/gatsby-plugin-utils/src/validate.ts:9]: Could you explain more exactly what this does?

File: packages/gatsby-plugin-utils/src/validate.ts
Comment (mxstbr) [reply packages/gatsby-plugin-utils/src/validate.ts:22]: Yes. `external` is explicitly the only way to add async rules.

File: packages/gatsby-plugin-utils/src/validate.ts
Comment (mxstbr) [reply packages/gatsby-plugin-utils/src/validate.ts:9]: I misunderstood how this works, thanks for the clarification. Here's an example:

```JS
const schema = Joi.object({
  str: Joi.string()
})

const options = validatePluginOptions(schema, {
  str: `bla`,
  notInSchema: true
});

// With stripUnknown: false
Error: 'notInSchema' is not allowed

// With stripUnknown: true
console.log(options) // => { str: `bla` }
```

I think we actually want the former now 😅

**Comment by mxstbr** [comment]
> One question I have is why a separate package?

I honestly can't remember the reason why, we for some reason said this should be a separate package. 🤔  @ascorbic do you remember by chance?

**Comment by ascorbic** [comment]
> > One question I have is why a separate package?
> 
> I honestly can't remember the reason why, we for some reason said this should be a separate package. 🤔 @ascorbic do you remember by chance?

Because we were planning to add tools that would allow plugin authors to validate their schemas, and to let users validate their configs in CI

**Comment by mxstbr** [comment]
> to let users validate their configs in CI

This would actually be really nice for plugin authors to use in their test setup. They can make sure options are validated/defaulted correctly based on what input/output they want!

File: packages/gatsby-plugin-utils/src/validate.ts
Comment (laurieontech) [reply packages/gatsby-plugin-utils/src/validate.ts:19]: Based on this I'm assuming we default to validating those rules if nothing is explicitly passed?

Comment (mxstbr) [reply packages/gatsby-plugin-utils/src/validate.ts:19]: Yes but we could default the other way as well

Comment (laurieontech) [reply packages/gatsby-plugin-utils/src/validate.ts:19]: Hard to know what we want at this moment. Seems reasonable to default true.

**Review by ascorbic — APPROVED** [review]
🎉

**Comment by MichaelDeBoey** [comment]
@mxstbr Should we add  `gatsby-plugin-utils` as a dependency of our plugin or should we wait until #27242 is merged instead? 🤔

Waiting for #27242 will cause a breaking change as the `gatsby` `peerDependency` will have to be upped to the version in which it's released I think?

**Comment by mxstbr** [comment]
> @mxstbr Should we add gatsby-plugin-utils as a dependency of our plugin or should we wait until #27242 is merged instead? 🤔

You're too quick! Give us a minute before using this, we'll be working on tooling for plugin authors based on this as well 👍

**Comment by ascorbic** [comment]
@mxstbr We probably need to make it clearer that this isn't ready for plugin authors to use quite yet. 
@MichaelDeBoey We will have some more treats for plugin authors coming soon. This is just the first step.

**Comment by MichaelDeBoey** [comment]
I'll have some patience then 😉

**Pull request merged**
